File: dogpack-developer/scalar/lib/hybrid/lib_poisson/unst/ConstructA_CG2.cpp
#include "mesh.h"
#include "FullMatrix.h"
#include "constants.h"

void ConstructA_CG2(const mesh& Mesh, FullMatrix& A)
{
  const int NumPhysElems = Mesh.get_NumPhysElems();
  const int NumBndNodes  = Mesh.get_SubNumBndNodes();
  const int Asize = Mesh.get_SubNumPhysNodes();

  assert_eq(Asize,A.get_NumRows());
  assert_eq(Asize,A.get_NumCols());
  
  dTensor1 A1(6);
  dTensor1 A2(6);
  dTensor1 A3(6);
  dTensor1 A4(6);
  dTensor1 A5(6);
  dTensor1 A6(6);

  A1.set(1, -oneninth     );
  A1.set(2,  4.0*oneninth );
  A1.set(3, -oneninth     );
  A1.set(4,  4.0*oneninth );
  A1.set(5,  4.0*oneninth );
  A1.set(6, -oneninth     );
  
  A2.set(1, -onethird     );
  A2.set(2,  0.0          );
  A2.set(3,  onethird     );
  A2.set(4, -4.0*onethird );
  A2.set(5,  4.0*onethird );
  A2.set(6,  0.0          );
  
  A3.set(1, -onethird     );
  A3.set(2, -4.0*onethird );
  A3.set(3,  0.0          );
  A3.set(4,  0.0          );
  A3.set(5,  4.0*onethird );
  A3.set(6,  onethird     );
  
  A4.set(1,  4.0          );
  A4.set(2, -4.0          );
  A4.set(3,  0.0          );
  A4.set(4, -4.0          );
  A4.set(5,  4.0          );
  A4.set(6,  0.0          );

  A5.set(1,  2.0          );
  A5.set(2, -4.0          );
  A5.set(3,  2.0          );
  A5.set(4,  0.0          );
  A5.set(5,  0.0          );
  A5.set(6,  0.0          );
  
  A6.set(1,  2.0          );
  A6.set(2,  0.0          );
  A6.set(3,  0.0          );
  A6.set(4, -4.0          );
  A6.set(5,  0.0          );
  A6.set(6,  2.0          );

  dTensor2 spts(3,2);
  spts.set(1,1,  1.0/3.0 );
  spts.set(1,2, -1.0/6.0 );
  
  spts.set(2,1, -1.0/6.0 );
  spts.set(2,2, -1.0/6.0 );
  
  spts.set(3,1, -1.0/6.0 );
  spts.set(3,2,  1.0/3.0 );
  
  dTensor1 wgts(3);
  wgts.set(1, 1.0/6.0 );
  wgts.set(2, 1.0/6.0 );
  wgts.set(3, 1.0/6.0 );
  
  // Loop over all elements in the mesh
  for (int i=1; i<=NumPhysElems; i++)
    {
      // Information for element i
      iTensor1 tt(6);
      for (int k=1; k<=6; k++)
	{  tt.set(k, Mesh.get_node_subs(i,k) );  }
      
      // Evaluate gradients of the Lagrange polynomials on Gauss quadrature points      
      dTensor2 gpx(6,3);
      dTensor2 gpy(6,3);
      
      for (int m=1; m<=3; m++)
	{
	  double  xi = spts.get(m,1);
	  double eta = spts.get(m,2);
	  
	  for (int k=1; k<=6; k++)
	    {
	      double gp_xi  = A2.get(k) + 2.0*A5.get(k)*xi + A4.get(k)*eta;
	      double gp_eta = A3.get(k) + A4.get(k)*xi + 2.0*A6.get(k)*eta;

	      gpx.set(k,m, Mesh.get_jmat(i,1,1)*gp_xi
		         + Mesh.get_jmat(i,1,2)*gp_eta );
	      gpy.set(k,m, Mesh.get_jmat(i,2,1)*gp_xi
		         + Mesh.get_jmat(i,2,2)*gp_eta );
	    }
	}

      // Entries of the stiffness matrix A
      double Area = Mesh.get_area_prim(i);
      for (int j=1; j<=6; j++)
	for (int k=1; k<=6; k++)
	  {
	    double tmp = A.get(tt.get(j),tt.get(k));
	    for (int m=1; m<=3; m++)
	      {
		tmp = tmp + 2.0*Area*wgts.get(m)*(gpx.get(j,m)*gpx.get(k,m)+gpy.get(j,m)*gpy.get(k,m));
	      }
	    A.set(tt.get(j),tt.get(k), tmp );
	  }
    }

  // Replace boundary node equations by Dirichlet boundary condition enforcement
  for (int i=1; i<=NumBndNodes; i++)
    {
      const int j=Mesh.get_sub_bnd_node(i);
      
      for (int k=1; k<=A.get_NumCols(); k++)
	{
	  A.set(j,k, 0.0 );	  
	}
      for (int k=1; k<=A.get_NumRows(); k++)
	{
	  A.set(k,j, 0.0 );
	}
      A.set(j,j, 1.0 );
    }

  // Get sparse structure representation
  A.Sparsify();
  
}

File: dogpack-mapped/scalar/lib/hybrid/lib_poisson/unst/matlab/ConstructA_CG2.m
function A = ConstructA_CG2(SubNumPhysNodes,NumPhysElems,...
                            SubNumBndNodes,sub_node,...
                            tnode,node_subs,area,sub_bnd_node,Jmat)
  
    A = sparse(SubNumPhysNodes,SubNumPhysNodes);
    
    A1 = zeros(6,1);
    A2 = zeros(6,1);
    A3 = zeros(6,1);
    A4 = zeros(6,1);
    A5 = zeros(6,1);
    A6 = zeros(6,1);
    
    oneninth = 1/9;
    onethird = 1/3;
    
    A1(1) = -oneninth;
    A1(2) =  4.0*oneninth;
    A1(3) = -oneninth;
    A1(4) =  4.0*oneninth;
    A1(5) =  4.0*oneninth;
    A1(6) = -oneninth;
  
    A2(1) = -onethird;
    A2(2) =  0.0;
    A2(3) =  onethird;
    A2(4) = -4.0*onethird;
    A2(5) =  4.0*onethird;
    A2(6) =  0.0;
  
    A3(1) = -onethird;
    A3(2) = -4.0*onethird;
    A3(3) =  0.0;
    A3(4) =  0.0;
    A3(5) =  4.0*onethird;
    A3(6) =  onethird;
    
    A4(1) =  4.0;
    A4(2) = -4.0;
    A4(3) =  0.0;
    A4(4) = -4.0;
    A4(5) =  4.0;
    A4(6) =  0.0;
    
    A5(1) =  2.0;
    A5(2) = -4.0;
    A5(3) =  2.0;
    A5(4) =  0.0;
    A5(5) =  0.0;
    A5(6) =  0.0;
    
    A6(1) =  2.0;
    A6(2) =  0.0;
    A6(3) =  0.0;
    A6(4) = -4.0;
    A6(5) =  0.0;
    A6(6) =  2.0;

    spts = zeros(3,2);
    spts(1,1) =  1.0/3.0;
    spts(1,2) = -1.0/6.0;
    spts(2,1) = -1.0/6.0;
    spts(2,2) = -1.0/6.0;
    spts(3,1) = -1.0/6.0;
    spts(3,2) =  1.0/3.0;
  
    wgts = zeros(3,1);
  
    wgts(1) = 1.0/6.0;
    wgts(2) = 1.0/6.0;
    wgts(3) = 1.0/6.0;
  
    % Loop over all elements in the mesh
    for i=1:NumPhysElems
    
        % Information for element i
        tt = node_subs(i,:);
                    
        % Evaluate gradients of the Lagrange polynomials on Gauss quadrature points      
        gpx = zeros(6,3);
        gpy = zeros(6,3);

        for m=1:3

            xi  = spts(m,1);
            eta = spts(m,2);
            
            for k=1:6

                gp_xi  = A2(k) + 2.0*A5(k)*xi + A4(k)*eta;
                gp_eta = A3(k) + A4(k)*xi + 2.0*A6(k)*eta;

                gpx(k,m) = Jmat(i,1,1)*gp_xi + Jmat(i,1,2)*gp_eta;
                gpy(k,m) = Jmat(i,2,1)*gp_xi + Jmat(i,2,2)*gp_eta;

            end
        end
        
        % Entries of the stiffness matrix A
        T = area(i);
        for j=1:6
            for k=1:6
                
                tmp = A(tt(j),tt(k));
                
                for m=1:3
                    tmp = tmp + 2.0*T*wgts(m)*(gpx(j,m)*gpx(k,m)+gpy(j,m)*gpy(k,m));
                end
                
                A(tt(j),tt(k)) = tmp;
                
            end
        end
        
    end
                

    % Replace boundary node equations by Dirichlet boundary condition enforcement
    for i=1:SubNumBndNodes
        
        j = sub_bnd_node(i);

        A(j,:) = 0;
        A(:,j) = 0;
        A(j,j) = 1;

    end

